examples/autoqasm/3_2_magic_state_distillation.ipynb

```diff
@@ -0,0 +1,447 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "6a4927a4-e1ed-4aea-9734-1993f5e741cd",
+   "metadata": {},
+   "source": [
+    "# Magic state distillation and gate teleportation\n",
+    "In fault tolerant quantum computing, Clifford gates have many desirable properties in stabilizer codes. For example, they can be implemented transversally and their noise is contained locally [1]. Therefore, there is a strong preference to include Clifford gates in a logical gateset. However, Clifford gates alone do not form a universal gateset. At least one non-Clifford gate is needed to perform universal quantum computing. Magic states are certain quantum states that enable universal fault-tolerant quantum computing with Clifford gates and preserve the desirable properties of the Clifford gates. With magic states, we can effectively apply non-Clifford gates using only Clifford gates, forming a universal gateset. \n",
+    "\n",
+    "In this notebook, we first demonstration, if we have access to magic states, how to implement a non-Clifford gate with only Clifford gates. This procedure is known as \"gate teleportation\". Because the preparation of magic state in this first demonstration is not fault-tolerant (i.e., it is not prepared under an error correction code), it is subjected to noise. In the second demonstration, we introduce a procedure to create logical magic states on the 5-qubit error correction code [2], called magic state distillation [1]. This procedure is based on post-selection, so the preparation of magic states has a finite probability of failing under this protocol. So, finally, we introduce a \"repeat until success\" protocol [3] that gaurantees every execution prepares magic state successfully. For all the protocols in this notebook, we use AutoQASM to program these procedures, demonstrating AutoQASM's ability to express the classical control flow needed for magic state distillation, showing how AutoQASM enables experimentations on fault-tolerant quantum comuting."
```

Comment:
I love this summary! Minor updates: typos and breaking into paragraphs

```suggestion
    "In this notebook, we first demonstrate how to implement a non-Clifford gate with only Clifford gates and magic states. This procedure is known as \"gate teleportation\". Because the preparation of the magic state in this first demonstration is not fault-tolerant (i.e., it is not prepared under an error correction code), it is subject to noise.
    
    In the second demonstration, we introduce a procedure to create logical magic states on the 5-qubit error correction code [2], called magic state distillation [1]. This procedure is based on post-selection, so the preparation of magic states has a finite probability of failing under this protocol.
    
    Finally, we introduce a \"repeat until success\" protocol [3] that guarantees every execution prepares a magic state successfully. For all the protocols in this notebook, we use AutoQASM to program these procedures, demonstrating AutoQASM's ability to express the classical control flow needed for magic state distillation and how AutoQASM enables early fault-tolerant quantum computing experiments."
```